server/README.md

```diff
@@ -0,0 +1,236 @@
+
+- Certificates
+- Blocks
+- Vlobs
+- Invitations
+
+### Certificates
+
+Any change done by a client that should be visible by the server is provided as a
+certificate (i.e. a certificate is an unencrypted document signed by the client's device).
+
+A very simple way to ensure causality and concurrency on certificates would be:
+
+- Have a single big lock in the server that must be taken before any attempt to
+  add a certificate (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  (handles causality).
+
+However this is pretty wasteful given most certificates are unrelated (e.g. certif A
+add a new user, certif B does a key rotation in a realm). So instead we reduce the
+size of the lock by introducing the concept of topic.
+
+Each type of certificate is part of one of the following topics:
+
+- `common`
+- `sequester`, which depends on `common`
+- `realm`, which depends on `common` and uses the realm ID as discriminant
+- `shamir_recovery`, which depends on `common` and use the ID of the user to recover as discriminant
+
+So the actual rule used in the server is:
+
+- Have lock in the server for each couple topic + discriminant
+- Take the lock for the topic + discriminant couple the certificate we want to add
+  belongs to AND to other topics our topic may depend on (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  in the topics we have locked (handles causality).
+
+> Note: Also organizations are totally isolated from each other on the server, so
+> in a way the organization ID is an aditional discriminant on each certificate topic.
+
+### Blocks
+
+Blocks are encrypted piece of data that belong to a given realm and cannot
+be modified.
+
+This makes concurrency & causality pretty simple:
+
+- Each block is considered independent, hence there is no need for lock dedicated to blocks.
+- A block is identified by a client-provided UUID, then there is no risk for double-creation
+  under concurrency (a trivial unicity check on insertion prevents this).
+- Block creation date must be more recent than the last certificate in `common` topic and
+  in the `realm` topic corresponding to the block's realm.
+
+> Notes:
+>
+> - It could be tempting to only lock the `realm` topic (given it's this topic that contains
+>   the certificate describing the role of a user in the realm). However a revoked user is also
+>   forbidden from adding data, hence making it mandatory to also lock the `common` topic here.
+> - Uploading blocks (and vlobs) are the most common task on the server, so we may want to
+>   uptomise it in the future to avoid having to lock two topics.
+>   A possible solution would be to have an optimistic locking approach: first read the
+>   last timestamp in the topics, then do the checks, then on the final data insertion check
+>   again the timestamp haven't changed (and if they have we re-do everything from the start).
```

Comment:
What would prevent a DoS if new certificates are continuously produced ?

Reply:
All API where lock is involved are authenticated, hence it is trivial for server administrator to freeze the users doing the DoS using the administration REST API.

Reply:
I understand that's trivial for server admin, but how that's possible on the SaaS ?

Reply:
I guess the whole point of SAAS is to not having to deal with things such as DoS (as it's the SAAS's administrators job) ;-)